SPARK-660: Add StorageLevel support in Python

aarondav opened the pull request:
It uses reflection... I am not proud of that fact, but it at least ensures
compatibility (sans refactoring of the StorageLevel stuff).

aarondav committed:
SPARK-660: Add StorageLevel support in Python

It uses reflection... I am not proud of that fact, but it at least ensures
compatibility (sans refactoring of the StorageLevel stuff).

AmplabJenkins commented:
Thank you for submitting this pull request.

All automated tests for this request have passed.

Refer to this link for build results: http://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/844/

pwendell commented:
Is there a more pythonic way to do enumerations? This is a neat trick to match the syntax of Scala's `StorageLevel.MEMORY_ONLY`, etc but if there is a more standard way of doing enumerations in python, I think it would be better to do that, even if we just hard coded the storage level options in the Python code. @JoshRosen should take a look though, he maintains the python API.

pwendell commented:
Hm it actually looks like every enumeration technique in python boils down to this same syntax ClassName.enumValue, so maybe this is in line with that...

jey commented:
Perhaps we should memoize the values on the Python side since they're constants?

aarondav committed:
Memoize StorageLevels read from JVM

AmplabJenkins commented:
Thank you for submitting this pull request.

All automated tests for this request have passed.

Refer to this link for build results: http://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/854/

mateiz commented on python/pyspark/context.py:
```diff
@@ -279,6 +279,27 @@ def setCheckpointDir(self, dirName, useExisting=False):
+            self.memoized[name] = storageLevel
+            return storageLevel
+        except:
+            print "Failed to find StorageLevel:", name
```
Let's not print stuff to stdout; throw an exception in this case

mateiz commented on python/pyspark/context.py:
```diff
@@ -279,6 +279,27 @@ def setCheckpointDir(self, dirName, useExisting=False):
+        self.sc = sc
+        self.memoized = {}
+
+    def __getattr__(self, name):
```
Instead of doing this and the reflection approach, I'd prefer to list the storage levels in Python. It's more repetitive, but it will be much easier to debug (e.g. no need to throw an exception below)

mateiz commented:
Thanks for adding this, Aaron. My comment though is that I'd prefer to list the storage levels in Python instead of using **getattr**.. that just creates a lot of complexity and error handling issues for a list of 6-7 constants. It's clever but not worth it for this use case.

aarondav committed:
Remove reflection, hard-code StorageLevels

The sc.StorageLevel -> StorageLevel pathway is a bit janky, but otherwise
the shell would have to call a private method of SparkContext. Having
StorageLevel available in sc also doesn't seem like the end of the world.
There may be a better solution, though.

As for creating the StorageLevel object itself, this seems to be the best
way in Python 2 for creating singleton, enum-like objects:
http://stackoverflow.com/questions/36932/how-can-i-represent-an-enum-in-python

AmplabJenkins commented:
Thank you for submitting this pull request.

All automated tests for this request have passed.

Refer to this link for build results: http://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/858/

mateiz commented on python/pyspark/shell.py:
```diff
@@ -29,6 +29,7 @@
 add_files = os.environ.get("ADD_FILES").split(',') if os.environ.get("ADD_FILES") != None else None
 
 sc = SparkContext(os.environ.get("MASTER", "local"), "PySparkShell", pyFiles=add_files)
+StorageLevel = sc.StorageLevel # alias StorageLevel to global scope
```
Unfortunately I don't think this is a good way to make this visible -- it will only be available in the shell. How about the following approach:
- Create an enum called StorageLevel in Python (e.g. using these options: http://stackoverflow.com/questions/36932/how-can-i-represent-an-enum-in-python)
- In RDD.persist(), translate the Python StorageLevel object, which is just an identifier, to a Java one, using a hashmap that you initialize inside SparkContext (when you're connected to Java).

aarondav committed:
Export StorageLevel and refactor

AmplabJenkins commented:
Thank you for submitting this pull request. 

Unfortunately, the automated tests for this request have failed.

Refer to this link for build results: http://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/864/

mateiz commented on python/pyspark/context.py:
```diff
@@ -47,6 +48,7 @@ class SparkContext(object):
     _active_spark_context = None
     _lock = Lock()
     _python_includes = None # zip and egg files that need to be added to PYTHONPATH
+    StorageLevel = None
```
This seems like it was left in unnecessarily

aarondav committed:
Whoopsy daisy

AmplabJenkins commented:
Thank you for submitting this pull request.

All automated tests for this request have passed.

Refer to this link for build results: http://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/867/

mateiz commented:
Looks good, thanks for putting this together!
